rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,572 @@
+---
+feature: feature_parameter_names
+start-date: 2024-01-10
+author: Dmitry Bogatov
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs has mechanisms to exports compile-time options of the upstream build
+system through the named parameters of derivation function, but the names of
+these parameters are inconsistent from package to package and from feature to
+feature.
+
+# Motivation
+[motivation]: #motivation
+
+Developer interface for building environment with specific features enabled and
+disabled is more complicated and and involves more "you just have to know it"
+compared to already existing systems.
+
+1. Knowing that package has a feature parameter (out of knowing the upstream)
+   does not mechanically translates to the name of feature parameter, it can be
+   named in multiple ways.
+
+2. Writing generic code is complicated. Even if list of packages all have
+   optional dependency on package `foo`, duck-typing like
+
+```
+map (p: p.override { with_foo = false; }) packages
+```
+
+won't work because of lack of uniformity, leading ugly code downstream.
+
+3. If user wants to globally disable support for optional feature or build
+   dependency as much as possible, he needs to set multiple feature parameters,
+   and he still can miss couple. E.g, it is not obvious whether the following
+   list is enough to cover everything X11-related:
+
+```
+withX
+withX11
+x11Support
+enableX
+```
```

Comment:
It sounds like this would be fixed with a [`config` option](https://github.com/NixOS/nixpkgs/blob/20d0954713ffa76a48507e90db6ad558f1238cf6/pkgs/top-level/config.nix), which is for flags that are generic/useful enough to apply to many packages.

Reply:
Overrides on individual packages become problematic (discussed in alternatives). Any way, we need some guidelines what exactly name we should to into `config`.

Reply:
Makes sense.

I think a convention to reuse the same `config` option name would be good, e.g. like this:

```nix
{
  config,
  withX11 ? config.withX11,
}:
...
```

Then we have consistency between per-package flags and global flags.

Reply:
I don't think we will need `config` for that. If you want global, just `overlays = [self: super: { with_x11 = false; }`